// alcalzone-shared/build/cjs/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var esm_exports = {};
__export(esm_exports, {
  arrays: () => arrays,
  async: () => async,
  comparable: () => comparable,
  deferredPromise: () => deferredPromise,
  helpers: () => helpers,
  math: () => math,
  objects: () => objects,
  typeguards: () => typeguards
});
module.exports = __toCommonJS(esm_exports);
var arrays = __toESM(require("./arrays/index.js"), 1);
var async = __toESM(require("./async/index.js"), 1);
var comparable = __toESM(require("./comparable/index.js"), 1);
var deferredPromise = __toESM(require("./deferred-promise/index.js"), 1);
var helpers = __toESM(require("./helpers/index.js"), 1);
var math = __toESM(require("./math/index.js"), 1);
var objects = __toESM(require("./objects/index.js"), 1);
var typeguards = __toESM(require("./typeguards/index.js"), 1);
