// @alcalzone/release-script-plugin-license/build/index.js
import { DefaultStages } from "@alcalzone/release-script-core";
import fs from "fs-extra";
import path from "path";
import glob from "tiny-glob";
class LicensePlugin {
    id = "license";
    stages = [DefaultStages.check];
    defineCLIOptions(yargs) {
        return yargs.options({
            license: {
                string: true,
                array: true,
                description: `Globs matching the license files to check`,
                default: ["{LICENSE,README}{,.md}"],
            },
        });
    }
    async executeStage(context, stage) {
        if (stage.id === "check") {
            const globs = context.argv.license;
            const files = [];
            for (const pattern of globs) {
                files.push(...(await glob(pattern, {
                    cwd: context.cwd,
                    dot: true,
                })));
            }
            for (const file of files) {
                const filePath = path.join(context.cwd, file);
                if (!(await fs.pathExists(filePath)))
                    continue;
                const fileContent = await fs.readFile(filePath, "utf8");
                const regex = /copyright\s*(\(c\))?\s*(?<range>(?:\d{4}\s*-\s*)?(?<current>\d{4}))/gi;
                let match;
                let latest;
                while ((match = regex.exec(fileContent))) {
                    if (!latest ||
                        parseInt(match.groups.current) > parseInt(latest.groups.current)) {
                        latest = match;
                    }
                }
                if (!latest)
                    continue;
                const latestYear = parseInt(latest.groups.current);
                if (latestYear < new Date().getFullYear()) {
                    context.cli.error(`File "${file}" contains an outdated copyright year: ${latest.groups.range}`);
                }
            }
        }
    }
}
export default LicensePlugin;
